src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -417,11 +420,14 @@ private Variant ReadVariant()
                 case EncodingKind.StringRef_1Byte:
                 case EncodingKind.StringRef_2Bytes:
                     return Variant.FromString(ReadStringValue(kind));
-                case EncodingKind.Object:
                 case EncodingKind.ObjectRef_4Bytes:
+                    return Variant.FromObject(_objectReferenceMap.GetValue(_reader.ReadInt32()));
```

Comment:
direct dispatch to each call.  instead of dispatching to ReadObject and then dispatching again.